### PR TITLE
Fix issue with PK lookup with lots of values

### PR DESCRIPTION
**Pull request by matriv**
When querying on a table using multiple columns which comprise the PRIMARY KEY, we use `EqualityExtractor` to extract the PK matches and build a `Get` plan instead of `Collect`. This operation can take lots of time when complex expressions or lots of values are tested against at least one of the PK columns. With: #16098 we introduced a MAX_ITERATIONS threshold to stop the process of extracting the pk matches, to avoid spending lots of time in this optimisation. But when the iteration threshold was reached we just exited the loop, and returned what was found so far, instead of returning `null` pk matches, and therefore falling back to a simple `Collect` operation.

Fixes: #16942
Relates: #16066

### CODE REVIEW

**Comment by proddata** [comment]
Could one force a refresh when falling back to a collect? And/or can we specify when this fallback happens? Otherwise this kinda seems to break with ...

> https://cratedb.com/docs/crate/reference/en/latest/concepts/storage-consistency.html#concept-consistency
> If a client updates a row and that row is looked up by its primary key after that update the changes will always be visible, since the information will be retrieved directly from the translog.

... or am I misinterpreting this?

**Comment by matriv** [comment]
> Could one force a refresh when falling back to a collect? And/or can we specify when this fallback happens? Otherwise this kinda seems to break with ...
> 
> > https://cratedb.com/docs/crate/reference/en/latest/concepts/storage-consistency.html#concept-consistency
> > If a client updates a row and that row is looked up by its primary key after that update the changes will always be visible, since the information will be retrieved directly from the translog.
> 
> ... or am I misinterpreting this?

If the query has complex expressions: https://github.com/crate/crate/pull/16098 or like this case lots of values with IN, then it will fallback to a Collect. This means that the user might not see the same records, as if it was a Get plan, without a manual (or automatic (refresh_interval)) refresh.

There is https://github.com/crate/crate/issues/16066 which tracks the situation, to improve the optimization in the future, and not have the need to fallback to Collect, but this is definitely a feature (as discussed within the team), and as a fix, the only thing we can do is to at least fallback to a Collect and not end up with a Get and less dockeys passed which leads for sure to wrong results.

**Comment by proddata** [comment]
> or like this case lots of values with IN

Can we specify that? What does lots of values imply? 5, 10, 100, 1000?
Or asked differently, is there currently a way to guarantee that I get the latest value (read-after-write-consistency) for selects?

File: server/src/main/java/io/crate/analyze/where/EqualityExtractor.java [hunk 1]
```diff
@@ -151,7 +151,7 @@ private EqMatches extractMatches(Collection<ColumnIdent> columns,
         for (List<EqProxy> proxies : cp) {
             // Protect against large queries, where the number of combinations to check grows large
             if (++iterations >= maxIterations()) {
-                break;
+                return new EqMatches(null, Set.of());
```
Comment (seut) [inline server/src/main/java/io/crate/analyze/where/EqualityExtractor.java:154]:
```suggestion
                // Fallback to lucene query (collect)
                return EqMatches.NONE;
```

**Comment by matriv** [comment]
> > or like this case lots of values with IN
> 
> Can we specify that? What does lots of values imply? 5, 10, 100, 1000? Or asked differently, is there currently a way to guarantee that I get the latest value (read-after-write-consistency) for selects?

Unfortunately, with the way the algorithm currently works in EqualityExtractor, we produce a cartesian product of the possible combinations of the where clause, and then we iterate over this product and try to extract dockeys, and leave the rest of the filtering to be done on top. So there is no absolute number that we can declare in the docs about this. it depends on how many are the pk cols, and how complex is the filtering.

**Comment by seut** [comment]
> Or asked differently, is there currently a way to guarantee that I get the latest value (read-after-write-consistency) for selects?

Yes, issue `REFRESH` before the select.

**Review by seut — COMMENTED** [review]
As @proddata mentioned, we should mention this exception also at https://cratedb.com/docs/crate/reference/en/latest/concepts/storage-consistency.html#concept-consistency.

**Comment by proddata** [comment]
>Yes, issue REFRESH before the select.
 But this breaks isolation?

Doesn't this also break the whole concept of Optimistic Concurrency Control https://cratedb.com/docs/crate/reference/en/latest/general/occ.html ?

File: docs/concepts/storage-consistency.rst [hunk 2]
```diff
@@ -176,6 +176,28 @@ translog.
   a replica shard is accessed in terms of consistency. Only the refresh of the
   ``IndexReader`` affects consistency.
 
+.. NOTE::
```
Comment (seut) [inline docs/concepts/storage-consistency.rst:179]: Instead of copying the content, I think it would be better to just link to the related content directly after the related claim at https://github.com/crate/crate/blob/master/docs/concepts/storage-consistency.rst?plain=1#L162.
E.g. 
`There is an exception to that when a lot of Primary Key values are detected inside a `WHERE` clause. See ... for details.`

**Comment by seut** [comment]
> Doesn't this also break the whole concept of Optimistic Concurrency Control https://cratedb.com/docs/crate/reference/en/latest/general/occ.html ?

Why? As `OCC` requires additional values to the PK (_seq + _primaryTerm), I 1) see no real use-case where one issue a statement affecting thousands of records 2) we document the exception when lot of PK values are used.

**Comment by seut** [comment]
@matriv Can you please also link to this exception at the `OCC` "Known limitations" section?

File: server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java [hunk 3]
```diff
@@ -396,7 +396,7 @@ public void test_pk_extraction_breaks_after_x_iterations() {
         EqualityExtractor extractor = new EqualityExtractor(normalizer) {
             @Override
             protected int maxIterations() {
-                return 1;
+                return 100;
```
Comment (seut) [inline server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java:399]: Why are these changes here needed?

Comment (matriv) [reply server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java:399]: the 100->1000 change is needed, because with 100 now we return null matches, but before we returned an incomplete list.
the 1 -> 100 is a safe number to demonstrate that even 100 iterations are not enough, this is not necessary, but thought it's nicer than 1 where one could assume anyway will never work since we join the `x = ? AND y = ?` 20 times. If you think is unnecessary I can revert this one.

Comment (seut) [reply server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java:399]: Ah I see. Probably 2 short comments would be good here.

**Review by seut — APPROVED** [review]

File: server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java (hunk 3 above)
Comment (matriv) [reply server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java:399]: thx, done!
